sw/device/silicon_creator/lib/cert/asn1.c

```diff
@@ -0,0 +1,181 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/silicon_creator/lib/cert/asn1.h"
+
+rom_error_t asn1_start(asn1_state_t *new_state, uint8_t *buffer, size_t size) {
+  // Make sure that the buffer is not too large to prevent overflows.
+  if (size > PTRDIFF_MAX)
+    return kErrorAsn1InvalidArgument;
+  new_state->buffer = buffer;
+  new_state->size = size;
+  new_state->offset = 0;
+  return kErrorOk;
+}
+
+rom_error_t asn1_finish(asn1_state_t *state, size_t *out_size) {
+  *out_size = state->offset;
+  state->buffer = 0;
+  state->size = 0;
+  state->offset = 0;
+  return kErrorOk;
+}
+
+rom_error_t asn1_push_byte(asn1_state_t *state, uint8_t byte) {
+  return asn1_push_bytes(state, &byte, 1);
+}
+
+rom_error_t asn1_push_bytes(asn1_state_t *state, const uint8_t *bytes,
+                            size_t size) {
+  // Make sure that the addition will not overflow
+  if (size > PTRDIFF_MAX)
+    return kErrorAsn1InvalidArgument;
+  if (state->offset + size > state->size)
```

Comment:
This test should suffice, if `size` is smaller than the certificate buffer we can trust it would be smaller than `PTRDIFF_MAX`?